Number_types/include/CGAL/boost_mp.h

```diff
@@ -607,6 +782,325 @@ namespace internal {
+
+          const Interval_nt<> quot = xn / xd;
+          CGAL_assertion(are_bounds_correct(quot.inf(), quot.sup(), x));
+          return std::make_pair(quot.inf(), quot.sup());
+        } */
+
+        // TODO: This is a temporary implementation and
+        // should be replaced by the default one. The default one fails:
+        // For some reason, CGAL::ldexp on Interval_nt returns
+        // 2.752961027411077506e-308 instead of denorm_min! We can work it around:
+
+        // See get_1ulp_interval() below:
+        // const auto res = CGAL::ldexp(intv, -static_cast<int>(shift)).pair();
+        // if (res.first == 0.0) {
+        //   CGAL_assertion(res.second != 0.0);
+        //   return std::make_pair(0.0, CGAL_IA_MIN_DOUBLE);
+        // }
+        // return res;
+
+        Interval_nt<false>
+        my_ldexp( const Interval_nt<false>& intv, const int e ) const {
+
+          CGAL_assertion(intv.inf() > 0.0);
+          CGAL_assertion(intv.sup() > 0.0);
+          const double scale = std::ldexp(1.0, e);
+          return Interval_nt<false> (
+            CGAL_NTS is_finite(scale) ?
+            scale * intv.inf() : CGAL_IA_MAX_DOUBLE,
+            scale == 0.0 ? CGAL_IA_MIN_DOUBLE : scale * intv.sup() );
+        }
+
+        std::pair<double, double> get_0ulp_interval(
+          const int64_t shift,
+          const boost::multiprecision::cpp_int& p ) const {
+
+          CGAL_assertion(p >= 0);
+          const uint64_t pp = static_cast<uint64_t>(p);
+          CGAL_assertion(pp >= 0);
+          const double pp_dbl = static_cast<double>(pp);
+          const Interval_nt<false> intv(pp_dbl, pp_dbl);
+          return my_ldexp(intv, -static_cast<int>(shift)).pair();
+        }
+
+        std::pair<double, double> get_1ulp_interval(
+          const int64_t shift,
+          const boost::multiprecision::cpp_int& p ) const {
+
+          CGAL_assertion(p >= 0);
+          const uint64_t pp = static_cast<uint64_t>(p);
+          const uint64_t qq = pp + 1;
+          CGAL_assertion(pp >= 0);
+          CGAL_assertion(qq > pp);
+          const double pp_dbl = static_cast<double>(pp);
+          const double qq_dbl = static_cast<double>(qq);
+          const Interval_nt<false> intv(pp_dbl, qq_dbl);
+          return my_ldexp(intv, -static_cast<int>(shift)).pair();
+        }
+
+        std::pair<double, double> operator()( Type x ) const {
+
+          CGAL_assertion(!CGAL::is_zero(x.den));
+          CGAL_assertion_code(const Type input = x);
+          double l = 0.0, u = 0.0;
+          if (CGAL::is_zero(x.num)) { // return [0.0, 0.0]
+            CGAL_assertion(are_bounds_correct(l, u, input));
+            return std::make_pair(l, u);
+          }
+          CGAL_assertion(!CGAL::is_zero(x.num));
+
+          // Handle signs.
+          bool change_sign = false;
+          const bool is_num_pos = CGAL::is_positive(x.num);
+          const bool is_den_pos = CGAL::is_positive(x.den);
```

Comment:
I think you mentioned that a negative denominator can only appear here because of a bug in a specific version of boost? If so, a comment and link to the bug would be helpful, I'd like to replace this with an assertion at some point in the future, if possible.

Reply:
Actually, it was not here. It was when constructing a cpp rational. Here, we handle the sign of a general rational number that can be either negative or positive.

Reply:
https://github.com/CGAL/cgal/pull/5937#discussion_r703599730

Reply:
My question was whether boost has an invariant that `denominator` always returns a positive number. But it doesn't seem documented, although some backends do provide this invariant, so I guess it would be ok for a rational backend not to provide it, and you are right to check the sign.